sonar-xml-plugin/src/main/java/org/sonar/plugins/xml/newchecks/NewlineCheck.java

```diff
@@ -0,0 +1,145 @@
+      checkChildrenLine(children, currentElement);
+      checkNextSiblingLine(currentElement);
+    }
+
+    children.forEach(this::visitNode);
+  }
+
+  private void checkChildrenLine(List<Node> children, Element currentElement) {
+    OutermostChildElements outermostChildElements = getOutermostChildElements(children);
+    XmlTextRange reportedRange = null;
+
+    if (outermostChildElements.first != null) {
+      XmlTextRange start = NewXmlFile.startLocation(currentElement);
+      XmlTextRange firstChildElementStart = NewXmlFile.startLocation(outermostChildElements.first);
+      if (firstChildElementStart.getStartLine() == start.getEndLine()) {
+        reportedRange = firstChildElementStart;
+      }
+    }
+
+    if (outermostChildElements.last != null) {
+      XmlTextRange end = NewXmlFile.endLocation(currentElement);
+      XmlTextRange lastChildElementRange = NewXmlFile.nodeLocation(outermostChildElements.last);
+      XmlTextRange lastChildElementEnd = NewXmlFile.endLocation(outermostChildElements.last);
+      if (lastChildElementEnd.getEndLine() == end.getStartLine()) {
+        boolean reportEntireElement = outermostChildElements.last.equals(outermostChildElements.first)
+          && reportedRange != null
+          && lastChildElementRange.getStartLine() == lastChildElementRange.getEndLine();
+        if (reportEntireElement) {
+          reportedRange = lastChildElementRange;
+        } else {
+          reportIssue(lastChildElementEnd, MESSAGE_END, Collections.emptyList());
+        }
+      }
+    }
+
+    if (reportedRange != null) {
+      reportIssue(reportedRange, MESSAGE_START, Collections.emptyList());
+    }
+  }
+
+  private void checkNextSiblingLine(Element node) {
+    XmlTextRange end = NewXmlFile.endLocation(node);
+    Element nextSiblingElement = getNextSiblingElement(node);
+    if (nextSiblingElement != null){
+      XmlTextRange nextSiblingElementStart = NewXmlFile.startLocation(nextSiblingElement);
+      if (nextSiblingElementStart.getStartLine() == end.getEndLine()) {
+        reportIssue(nextSiblingElementStart, MESSAGE_START, Collections.emptyList());
+      }
+    }
+  }
+
+  @Nullable
+  private static Element getNextSiblingElement(Node node) {
+    Node nextSibling = node.getNextSibling();
+    if (nextSibling == null) {
+      return null;
+    }
+    if (nextSibling.getNodeType() == Node.ELEMENT_NODE) {
+      return (Element) nextSibling;
+    }
+
+    return getNextSiblingElement(nextSibling);
+  }
+
+  private static OutermostChildElements getOutermostChildElements(List<Node> children) {
+    Element firstChildElement = null;
+    Element lastChildElement = null;
+
+    for (Node child : children) {
+      if (child.getNodeType() == Node.ELEMENT_NODE) {
+        if (firstChildElement == null) {
+          firstChildElement = (Element) child;
+        }
+        lastChildElement = (Element) child;
+      }
+    }
+
+    return new OutermostChildElements(firstChildElement, lastChildElement);
```

Comment:
It seems to me that the only cases where first and last will be null is when there is no children being ELEMENT. In such case, we are not reporting anything through `checkChildrenLine`.

Instead of having `@Nullable` first/last I would simply not create the object if both are null, and return null.